pkg/controller/ipam/metrics.go

```diff
@@ -0,0 +1,150 @@
+	"antrea.io/antrea/pkg/client/clientset/versioned"
+	crdinformers "antrea.io/antrea/pkg/client/informers/externalversions"
+	"antrea.io/antrea/pkg/client/informers/externalversions/crd/v1alpha2"
+	crdlisters "antrea.io/antrea/pkg/client/listers/crd/v1alpha2"
+	"antrea.io/antrea/pkg/ipam/poolallocator"
+)
+
+const (
+	controllerName        = "AntreaIPAMController"
+	metricNamespaceAntrea = "antrea"
+	metricSubsystemIPAM   = "ipam"
+)
+
+type AntreaIPAMMetricsHandler struct {
+	crdClient      versioned.Interface
+	ipPoolLister   crdlisters.IPPoolLister
+	ipPoolInformer v1alpha2.IPPoolInformer
+}
+
+var (
+	ipPoolUsageTotal = metrics.NewGaugeVec(
+		&metrics.GaugeOpts{
+			Namespace:      metricNamespaceAntrea,
+			Subsystem:      metricSubsystemIPAM,
+			Name:           "ippool_usage_total",
+			Help:           "Total number of available IP addresses for the address pool.",
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"pool_name"},
+	)
+	ipPoolUsageUsed = metrics.NewGaugeVec(
+		&metrics.GaugeOpts{
+			Namespace:      metricNamespaceAntrea,
+			Subsystem:      metricSubsystemIPAM,
+			Name:           "ippool_usage_used",
+			Help:           "Number of allocated IP addresses for the address pool.",
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"pool_name"},
+	)
+)
+
+func NewAntreaIPAMMetricsHandler(crdClient versioned.Interface, crdInformerFactory crdinformers.SharedInformerFactory) *AntreaIPAMMetricsHandler {
+	legacyregistry.MustRegister(ipPoolUsageTotal)
+	legacyregistry.MustRegister(ipPoolUsageUsed)
+
+	ipPoolInformer := crdInformerFactory.Crd().V1alpha2().IPPools()
+	mh := &AntreaIPAMMetricsHandler{
+		crdClient:      crdClient,
+		ipPoolInformer: ipPoolInformer,
+		ipPoolLister:   ipPoolInformer.Lister()}
+	return mh
+}
+
+func (c *AntreaIPAMMetricsHandler) Run(stopCh <-chan struct{}) {
+	c.ipPoolInformer.Informer().AddEventHandlerWithResyncPeriod(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.createHandler,
+		UpdateFunc: c.updateHandler,
+		DeleteFunc: c.deleteHandler,
+	}, 0)
+
+	if !cache.WaitForNamedCacheSync(controllerName, stopCh, c.ipPoolInformer.Informer().HasSynced) {
+		return
+	}
+	<-stopCh
+}
+
+func (c *AntreaIPAMMetricsHandler) createHandler(obj interface{}) {
+	c.updateIPPoolCounters(obj)
+}
+
+func (c *AntreaIPAMMetricsHandler) updateHandler(oldObj, newObj interface{}) {
+	c.updateIPPoolCounters(newObj)
+}
+
+func (c *AntreaIPAMMetricsHandler) deleteHandler(obj interface{}) {
+	ipPool := obj.(*crdv1a2.IPPool)
+
+	if !ipPoolUsageTotal.DeleteLabelValues(ipPool.Name) {
+		klog.Warningf("Deletion of IPPool %s ippool_usage_total metric failed", ipPool.Name)
+	}
+	if !ipPoolUsageUsed.DeleteLabelValues(ipPool.Name) {
+		klog.Warningf("Deletion of IPPool %s ippool_usage_total metric failed", ipPool.Name)
+	}
+}
+
+func (c *AntreaIPAMMetricsHandler) updateIPPoolCounters(obj interface{}) {
+	ipPool := obj.(*crdv1a2.IPPool)
+
+	allocator, err := poolallocator.NewIPPoolAllocator(ipPool.Name, c.crdClient)
```

Comment:
This will need to change a bit once https://github.com/antrea-io/antrea/pull/3046 is merged (pool informer is already included in this class so not a big deal)

Reply:
So maybe it's better if I rebase on top of #3046